sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1045 @@
+        }
+      }
+    }
+
+    /**
+     * Throws exception if you set different InputFormat key or value class than InputFormat's
+     * actual key or value class. If you set incorrect classes then, it may result in an error like
+     * "unexpected extra bytes after decoding" while the decoding process happens. Hence this
+     * validation is required.
+     */
+    private void validateUserInputForKeyAndValue() throws IOException, InterruptedException {
+      ParameterizedType genericClassType = determineGenericType();
+      boolean isCorrectKeyClassSet = validateClasses(
+          genericClassType.getActualTypeArguments()[0].getTypeName(),
+          keyCoder,
+          "key.class");
+      boolean isCorrectValueClassSet = validateClasses(
+          genericClassType.getActualTypeArguments()[1].getTypeName(),
+          valueCoder,
+          "value.class");
+      if (!isCorrectKeyClassSet) {
+        Class<?> actualClass = conf.getHadoopConfiguration().getClass("key.class", Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_KEY_CLASS_ERROR_MSG,
+            getExpectedKeyClass().getName(),
+            actualClass.getName()));
+      }
+      if (!isCorrectValueClassSet) {
+        Class<?> actualClass = conf.getHadoopConfiguration().getClass("value.class", Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_VALUE_CLASS_ERROR_MSG,
+            getExpectedValueClass().getName(),
+            actualClass.getName()));
+      }
+    }
+
+    /**
+     * Returns true if key/value class set by the user is compatible with the key/value class of a
+     * pair returned by RecordReader. User provided key and value classes are validated against
+     * parameterized type of InputFormat. If parameterized type has any type parameter such as T,
+     * K, V, etc then validation is done by encoding and decoding first pair returned by
+     * RecordReader.
+     */
+    private <T> boolean validateClasses(
+        String inputFormatGenericClassName,
+        Coder<T> coder,
+        String property) throws IOException, InterruptedException {
+      Class<?> inputClass = null;
+      try {
+        inputClass = Class.forName(inputFormatGenericClassName);
+      } catch (Exception e) {
+        /*
+         * Given inputFormatGenericClassName is a type parameter i.e. T, K, V, etc. In such cases
+         * class validation for user provided input key/value will not work correctly. Therefore the
+         * need to validate key/value classes by encoding and decoding key/value object with the
+         * given coder.
+         */
+        return validateClassUsingCoder(property, coder);
+      }
+      /*
+       * Validates key/value class with InputFormat's parameterized type.
+       */
+      if (inputClass != null) {
```

Comment:
You are right. This situation might not occur though. Code has been updated with some new changes which addresses this issue.